src/java.base/share/classes/java/lang/ClassLoader.java

```diff
@@ -1941,25 +1941,21 @@ public static ClassLoader getPlatformClassLoader() {
      */
     @CallerSensitive
     public static ClassLoader getSystemClassLoader() {
-        switch (VM.initLevel()) {
-            case 0:
-            case 1:
-            case 2:
-                // the system class loader is the built-in app class loader during startup
-                return getBuiltinAppClassLoader();
-            case 3:
-                String msg = "getSystemClassLoader cannot be called during the system class loader instantiation";
-                throw new IllegalStateException(msg);
-            default:
-                // system fully initialized
-                assert VM.isBooted() && scl != null;
-                @SuppressWarnings("removal")
-                SecurityManager sm = System.getSecurityManager();
-                if (sm != null) {
-                    checkClassLoaderPermission(scl, Reflection.getCallerClass());
-                }
-                return scl;
+        if (!VM.initLevelReached(VM.SYSTEM_LOADER_INITIALIZING)) {
+            // the system class loader is the built-in app class loader during startup
+            return getBuiltinAppClassLoader();
+        } else if (!VM.isBooted()) {
+            String msg = "getSystemClassLoader cannot be called during the system class loader instantiation";
+            throw new IllegalStateException(msg);
```

Comment:
This looks okay, but failing if VM.initLevel() == VM.SYSTEM_LOADER_INITIALIZING might be clearer.